README.md
## JDGitCodeTree

Browser extension (Chrome, Firefox, Opera and Safari) to show a code tree on git.jd.com.

fork from [https://github.com/buunguyen/octotree](https://github.com/buunguyen/octotree)




HISTORY.md
### v1.0
* First realese


src/octotree.js
$(document).ready(() => {
  const store = new Storage()

  parallel(Object.keys(STORE), setDefault, loadExtension)

  function setDefault(key, cb) {
    const storeKey = STORE[key]
    store.get(storeKey, (val) => {
      store.set(storeKey, val == null ? DEFAULTS[key] : val, cb)
    })
  }

  function createAdapter() {
    const normalizeUrl = (url) => url.replace(/(.*?:\/\/[^/]+)(.*)/, '$1')

    const githubUrls = store.get(STORE.GHEURLS).split(/\n/)
      .map(normalizeUrl)
      .concat('https://github.com')

    const bitbucketUrls = ['https://bitbucket.org']
    const jdgitUrls = [
      'https://git.jd.com','http://git.jd.com'
    ]
    const currentUrl = `${location.protocol}//${location.host}`

    if (~jdgitUrls.indexOf(currentUrl)) {
      return new JDGit(store)
    } else if (~githubUrls.indexOf(currentUrl)) {
      return new GitHub(store)
    } else if (~bitbucketUrls.indexOf(currentUrl)) {
      return new Bitbucket(store)
    }
  }

  function loadExtension() {
    const $html = $('html')
    const $document = $(document)
    const $dom = $(TEMPLATE)
    const $sidebar = $dom.find('.octotree_sidebar')
    const $toggler = $sidebar.find('.octotree_toggle')
    const $views = $sidebar.find('.octotree_view')
    const adapter = createAdapter()
    const treeView = new TreeView($dom, store, adapter)
    const optsView = new OptionsView($dom, store)
    const helpPopup = new HelpPopup($dom, store)
    const errorView = new ErrorView($dom, store)
    let currRepo = false
    let hasError = false

    $html.addClass(ADDON_CLASS)

    $(window).resize((event) => {
      if (event.target === window) layoutChanged()
    })

    $toggler.click(toggleSidebarAndSave)
    key.filter = () => $toggler.is(':visible')
    key(store.get(STORE.HOTKEYS), toggleSidebarAndSave)

    const views = [treeView, errorView, optsView]
    views.forEach((view) => {
      $(view)
        .on(EVENT.VIEW_READY, function (event) {
          if (this !== optsView) {
            $document.trigger(EVENT.REQ_END)
          }
          showView(this.$view)
        })
        .on(EVENT.VIEW_CLOSE, () => showView(hasError ? errorView.$view : treeView.$view))
        .on(EVENT.OPTS_CHANGE, optionsChanged)
        .on(EVENT.FETCH_ERROR, (event, err) => showError(err))
    })

    $document
      .on(EVENT.REQ_START, () => $toggler.addClass('octotree_loading'))
      .on(EVENT.REQ_END, () => $toggler.removeClass('octotree_loading'))
      .on(EVENT.LAYOUT_CHANGE, layoutChanged)
      .on(EVENT.TOGGLE, layoutChanged)
      .on(EVENT.LOC_CHANGE, () => tryLoadRepo())

    $sidebar
      .width(parseInt(store.get(STORE.WIDTH)))
      .resize(() => layoutChanged(true))
      .appendTo($('body'))

    adapter.init($sidebar)
    return tryLoadRepo()

    function optionsChanged(event, changes) {
      let reload = false

      Object.keys(changes).forEach((storeKey) => {
        const value = changes[storeKey]

        switch (storeKey) {
          case STORE.TOKEN:
          case STORE.LOADALL:
          case STORE.ICONS:
            reload = true
            break
          case STORE.HOTKEYS:
            key.unbind(value[0])
            key(value[1], toggleSidebar)
            break
        }
      })

      if (reload) {
        tryLoadRepo(true)
      }
    }

    function tryLoadRepo(reload) {
      hasError = false
      const remember = store.get(STORE.REMEMBER)
      const shown = store.get(STORE.SHOWN)
      const token = store.get(STORE.TOKEN)

      adapter.getRepoFromPath(currRepo, token, (err, repo) => {
        if (err) {
          showError(err)
        }
        else if (repo) {
          $toggler.show()

          if (remember && shown) {
            toggleSidebar(true)
          }

          if (isSidebarVisible()) {
            const replacer = ['username', 'reponame', 'branch', 'pullNumber']
            const repoChanged = JSON.stringify(repo, replacer) !== JSON.stringify(currRepo, replacer)
            if (repoChanged || reload === true) {
              $document.trigger(EVENT.REQ_START)
              currRepo = repo
              treeView.show(repo, token)
            }
            else {
              treeView.syncSelection()
            }
          }
        }
        else {
          $toggler.hide()
          toggleSidebar(false)
        }
        helpPopup.init()
        layoutChanged()
      })
    }

    function showView(view) {
      $views.removeClass('current')
      view.addClass('current')
    }

    function showError(err) {
      hasError = true
      errorView.show(err)
    }

    function toggleSidebarAndSave() {
      store.set(STORE.SHOWN, !isSidebarVisible(), () => {
        toggleSidebar()
        if (isSidebarVisible()) {
          tryLoadRepo()
        }
      })
    }

    function toggleSidebar(visibility) {
      if (visibility !== undefined) {
        if (isSidebarVisible() === visibility) return
        toggleSidebar()
      }
      else {
        $html.toggleClass(SHOW_CLASS)
        $document.trigger(EVENT.TOGGLE, isSidebarVisible())
      }
    }

    function layoutChanged(save = false) {
      const width = $sidebar.outerWidth()
      adapter.updateLayout(isTogglerVisible(), isSidebarVisible(), width)
      if (save === true) {
        store.set(STORE.WIDTH, width)
      }
    }

    function isSidebarVisible() {
      return $html.hasClass(SHOW_CLASS)
    }

    function isTogglerVisible() {
      return $toggler.is(':visible')
    }
  }
})


gulpfile.babel.js
import gulp from 'gulp'
import fs from 'fs'
import os from 'os'
import path from 'path'
import { merge } from 'event-stream'
import map from 'map-stream'
import { spawn } from 'child_process'
const $ = require('gulp-load-plugins')()

// Tasks
gulp.task('clean', () => {
  return pipe('./tmp', $.clean())
})

gulp.task('build', (cb) => {
  $.runSequence('clean', 'styles', 'chrome', 'opera', 'safari', 'firefox', cb)
})

gulp.task('default', ['build'], () => {
  gulp.watch(['./libs/**/*', './src/**/*', './package.json'], ['default'])
})

gulp.task('dist', ['build'], (cb) => {
  $.runSequence('firefox:zip', 'chrome:zip', 'chrome:crx', 'opera:nex', cb)
})

gulp.task('test', ['build'], (cb) => {
  const ps = spawn(
    './node_modules/.bin/mocha',
    ['--harmony', '--reporter', 'spec', '--bail', '--recursive', '--timeout', '-1']
  )
  ps.stdout.pipe(process.stdout);
  ps.stderr.pipe(process.stderr);
  ps.on('close', cb)
})

gulp.task('styles', () => {
  return pipe(
    './src/styles/octotree.less',
    $.plumber(),
    $.less({ relativeUrls: true }),
    $.autoprefixer({ cascade: true }),
    './tmp'
  )
})

gulp.task('lib:ondemand', (cb) => {
  const dir = './libs/ondemand'
  const code = fs.readdirSync(dir).map(file => {
    return `window['${file}'] = function () {
    ${fs.readFileSync(path.join(dir, file))}
  };\n`
  }).join('')

  fs.writeFileSync('./tmp/ondemand.js', code)

  cb()
})

// Chrome
gulp.task('chrome:template', () => {
  return buildTemplate({SUPPORT_FILE_ICONS: true, SUPPORT_GHE: true})
})

gulp.task('chrome:js', ['chrome:template', 'lib:ondemand'], () => {
  return buildJs(['./src/config/chrome/overrides.js'], {SUPPORT_FILE_ICONS: true, SUPPORT_GHE: true})
})

gulp.task('chrome', ['chrome:js'], () => {
  return merge(
    pipe('./icons/**/*', './tmp/chrome/icons'),
    pipe(['./libs/**/*', '!./libs/ondemand{,/**}', './tmp/octotree.*', './tmp/ondemand.js'], './tmp/chrome/'),
    pipe('./libs/file-icons.css', $.replace('../fonts', 'chrome-extension://__MSG_@@extension_id__/fonts'), './tmp/chrome/'),
    pipe('./src/config/chrome/background.js', $.babel(), './tmp/chrome/'),
    pipe('./src/config/chrome/manifest.json', $.replace('$VERSION', getVersion()), './tmp/chrome/')
  )
})

gulp.task('chrome:zip', () => {
  return pipe('./tmp/chrome/**/*', $.zip('chrome.zip'), './dist')
})

gulp.task('chrome:crx', () => {
  // This will package the crx using a private key.
  // For the convenience of people who want to build locally without having to
  // manage their own Chrome key, this code will use the bundled test key if
  // a real key is not found in ~/.ssh.
  const real = path.join(os.homedir() + '.ssh/chrome.pem')
  const test = './chrome_test_key.pem'
  const privateKey = fs.existsSync(real) ? fs.readFileSync(real) : fs.readFileSync(test)

  return pipe('./tmp/chrome', $.crxPack({
    privateKey: privateKey,
    filename: 'chrome.crx'
  }), './dist')
})

// Opera
gulp.task('opera', ['chrome'], () => {
  return pipe('./tmp/chrome/**/*', './tmp/opera')
})

gulp.task('opera:nex', () => {
  return pipe('./dist/chrome.crx', $.rename('opera.nex'), './dist')
})

// Firefox
gulp.task('firefox:template', () => {
  return buildTemplate({SUPPORT_FILE_ICONS: true})
})

gulp.task('firefox:js', ['firefox:template', 'lib:ondemand'], () => {
  return buildJs([], {SUPPORT_FILE_ICONS: true})
})

gulp.task('firefox', ['firefox:js'], () => {
  return merge(
    pipe('./icons/**/*', './tmp/firefox/icons'),
    pipe(['./libs/**/*', '!./libs/ondemand{,/**}', './tmp/octotree.*', './tmp/ondemand.js'], './tmp/firefox'),
    pipe('./libs/file-icons.css', $.replace('../fonts', 'moz-extension://__MSG_@@extension_id__/fonts'), './tmp/firefox/'),
    pipe('./src/config/firefox/background.js', $.babel(), './tmp/firefox/'),
    pipe('./src/config/firefox/manifest.json', $.replace('$VERSION', getVersion()), './tmp/firefox')
  )
})

gulp.task('firefox:zip', () => {
  return pipe('./tmp/firefox/**/*', $.zip('firefox.zip'), './dist')
})

// Safari
gulp.task('safari:template', () => {
  return buildTemplate({SUPPORT_FILE_ICONS: true})
})

gulp.task('safari:js', ['safari:template', 'lib:ondemand'], () => {
  return buildJs([], {SUPPORT_FILE_ICONS: true})
})

gulp.task('safari', ['safari:js'], () => {
  return merge(
    pipe('./icons/icon64.png', $.rename('Icon-64.png'), './tmp/safari/octotree.safariextension'),
    pipe(
      ['./libs/**/*', '!./libs/ondemand{,/**}', './tmp/octotree.*', './tmp/ondemand.js'],
      './tmp/safari/octotree.safariextension/'
    ),
    pipe('./libs/file-icons.css', $.replace('../fonts', 'fonts'), './tmp/safari/octotree.safariextension/'),
    pipe('./src/config/safari/Info.plist', $.replace('$VERSION', getVersion()), './tmp/safari/octotree.safariextension')
  )
})

// Helpers
function pipe(src, ...transforms) {
  return transforms.reduce((stream, transform) => {
    const isDest = typeof transform === 'string'
    return stream.pipe(isDest ? gulp.dest(transform) : transform)
  }, gulp.src(src))
}

function html2js(template) {
  return map(escape)

  function escape(file, cb) {
    const path = $.util.replaceExtension(file.path, '.js')
    const content = file.contents.toString()
    const escaped = content
      .replace(/\\/g, "\\\\")
      .replace(/'/g, "\\'")
      .replace(/\r?\n/g, "\\n' +\n    '")
    const body = template.replace('$$', escaped)

    file.path = path
    file.contents = new Buffer(body)
    cb(null, file)
  }
}

function buildJs(overrides, ctx) {
  const src = [
    './tmp/template.js',
    './src/constants.js',
    './src/adapters/adapter.js',
    './src/adapters/pjax.js',
    './src/adapters/bitbucket.js',
    './src/adapters/github.js',
    './src/adapters/jdgit.js',
    './src/view.help.js',
    './src/view.error.js',
    './src/view.tree.js',
    './src/view.options.js',
    './src/util.location.js',
    './src/util.module.js',
    './src/util.async.js',
    './src/util.storage.js'
  ].concat(overrides)
   .concat('./src/octotree.js')

  return pipe(
    src,
    $.babel(),
    $.concat('octotree.js'),
    $.preprocess({context: ctx}),
    './tmp'
  )
}

function buildTemplate(ctx) {
  const LOTS_OF_SPACES = new Array(500).join(' ')

  return pipe(
    './src/template.html',
    $.preprocess({ context: ctx }),
    $.replace('__SPACES__', LOTS_OF_SPACES),
    html2js('const TEMPLATE = \'$$\''),
    './tmp'
  )
}

function getVersion() {
  delete require.cache[require.resolve('./package.json')]
  return require('./package.json').version
}


src/adapters/pjax.js
class PjaxAdapter extends Adapter {
  constructor(store) {
    super(['jquery.pjax.js'], store)

    $(document)
      .on('pjax:start', () => $(document).trigger(EVENT.REQ_START))
      .on('pjax:end', () => $(document).trigger(EVENT.REQ_END))
      .on('pjax:timeout', (e) => e.preventDefault())
  }

  // @override
  // @param {Object} opts - {pjaxContainer: the specified pjax container}
  // @api public
  init($sidebar, opts) {
    super.init($sidebar)

    opts = opts || {}
    const pjaxContainer = opts.pjaxContainer

    if (!window.MutationObserver) return

    // Some host switch pages using pjax. This observer detects if the pjax container
    // has been updated with new contents and trigger layout.
    const pageChangeObserver = new window.MutationObserver(() => {
      // Trigger location change, can't just relayout as Octotree might need to
      // hide/show depending on whether the current page is a code page or not.
      return $(document).trigger(EVENT.LOC_CHANGE)
    })

    if (pjaxContainer) {
      pageChangeObserver.observe(pjaxContainer, {
        childList: true,
      })
    }
    else { // Fall back if DOM has been changed
      let firstLoad = true, href, hash

      function detectLocChange() {
        if (location.href !== href || location.hash !== hash) {
          href = location.href
          hash = location.hash

          // If this is the first time this is called, no need to notify change as
          // Octotree does its own initialization after loading options.
          if (firstLoad) {
            firstLoad = false
          }
          else {
            setTimeout(() => {
              $(document).trigger(EVENT.LOC_CHANGE)
            }, 300) // Wait a bit for pjax DOM change
          }
        }
        setTimeout(detectLocChange, 200)
      }

      detectLocChange()
    }
  }

  // @override
  // @param {Object} opts - {$pjax_container: jQuery object}
  // @api public
  selectFile(path, opts) {
    opts = opts || {}
    const $pjaxContainer = opts.$pjaxContainer
    const fragment = opts.fragment

    // if we're on the same page and just navigating to a different anchor
    // don't bother fetching the page with pjax
    const pathWithoutAnchor = path.replace(/#.*$/, '')
    const isSamePage = location.pathname === pathWithoutAnchor
    const loadWithPjax = $pjaxContainer.length && !isSamePage

    if (loadWithPjax) {
      $.pjax({
        // needs full path for pjax to work with Firefox as per cross-domain-content setting
        url: location.protocol + '//' + location.host + path,
        container: '#' + $pjaxContainer.attr('id'),
        fragment: fragment,
        timeout: 0 // global timeout doesn't seem to work, use this instead
      })
    }
    else {
      super.selectFile(path)
    }
  }
}


src/adapters/jdgit.js
const OSC_RESERVED_USER_NAMES = ["explore", "repositories", "popular", "enterprises", "gists", "dashboard", "languages", "search", "links", "invite", "profile", "organizations", "notifications", "login", "signup", "oauth"]
const OSC_RESERVED_REPO_NAMES = ["admin", "dashboard", "groups", "help", "profile", "projects", "search", "codes", "fork_project", "fork_code"]
const OSC_404_SEL = '#parallax_wrapper'
const OSC_PJAX_CONTAINER_SEL = '#git-project-content'
const OSC_CONTAINERS = '#git-header-nav'
const OSC_RAW_CONTENT = 'body > pre'

class JDGit extends PjaxAdapter {

  constructor(store) {
    super(store)
  }

  // @override
  init($sidebar) {
    const pjaxContainer = $(OSC_PJAX_CONTAINER_SEL)[0]
    super.init($sidebar, { 'pjaxContainer': pjaxContainer })

    // Fix #151 by detecting when page layout is updated.
    // In this case, split-diff page has a wider layout, so need to recompute margin.
    // Note that couldn't do this in response to URL change, since new DOM via pjax might not be ready.
    const diffModeObserver = new window.MutationObserver((mutations) => {
      mutations.forEach((mutation) => {
        if (~mutation.oldValue.indexOf('split-diff') ||
            ~mutation.target.className.indexOf('split-diff')) {
          return $(document).trigger(EVENT.LAYOUT_CHANGE)
        }
      })
    })

    diffModeObserver.observe(document.body, {
      attributes: true,
      attributeFilter: ['class'],
      attributeOldValue: true
    })
  }

  // @override
  _getCssClass() {
    return 'octotree_oschina_sidebar'
  }

  // @override
  canLoadEntireTree() {
    return true
  }

  // @override
  getCreateTokenUrl() {
    return `https://gitee.com/api/v5/swagger`
  }

  // @override
  updateLayout(togglerVisible, sidebarVisible, sidebarWidth) {
    const SPACING = 232
    const $containers = $(OSC_CONTAINERS)
    const autoMarginLeft = ($(document).width() - $containers.width()) / 2
    const WIDTH = $(document).width() - SPACING
    const shouldPushLeft = sidebarVisible && (autoMarginLeft <= sidebarWidth + SPACING)

    $('html').css('margin-left', shouldPushLeft ? sidebarWidth : '')
    $containers.css('margin-left', shouldPushLeft ? SPACING : '')
    $containers.css('width', shouldPushLeft ? WIDTH : '')
    // $(".ui.right.floated.horizontal.list").css('margin-right', shouldPushLeft ? 210 : '')
    $(".git-project-download-panel").css('margin-right', shouldPushLeft ? 240 : '')
  }

  // @override
  getRepoFromPath(currentRepo, token, cb) {
    const showInNonCodePage = this.store.get(STORE.NONCODE)

    // 404 page, skip
    if ($(OSC_404_SEL).length) {
      return cb()
    }

    // Skip raw page
    if ($(OSC_RAW_CONTENT).length) {
      return cb()
    }

    // (username)/(reponame)[/(type)]
    const match = window.location.pathname.match(/([^\/]+)\/([^\/]+)(?:\/([^\/]+))?/)
    if (!match) {
      return cb()
    }

    const username = match[1]
    const reponame = match[2]

    // Not a repository, skip
    if (~OSC_RESERVED_USER_NAMES.indexOf(username) ||
        ~OSC_RESERVED_REPO_NAMES.indexOf(reponame)) {
      return cb()
    }

    // Skip non-code page unless showInNonCodePage is true
    if (!showInNonCodePage && match[3] && !~['tree', 'blob'].indexOf(match[3])) {
      return cb()
    }

    // Get branch by inspecting page, quite fragile so provide multiple fallbacks
    const branch =
    // Code page
    $('.project-refs-form .dropdown-toggle-text ').text().trim() ||
    // Pull requests page
    ($('.commit-ref.base-ref').attr('title') || ':').match(/:(.*)/)[1] ||
    // Reuse last selected branch if exist
    (currentRepo.username === username && currentRepo.reponame === reponame && currentRepo.branch) ||
    // Get default branch from cache
    this._defaultBranch[username + '/' + reponame]

    // Still no luck, get default branch for real
    const repo = { username: username, reponame: reponame, branch: branch }

    if (repo.branch) {
      cb(null, repo)
    } else {
      this._get(null, { repo, token }, (err, data) => {
        if (err) return cb(err)
        repo.branch = this._defaultBranch[username + '/' + reponame] = data.default_branch || 'master'
        cb(null, repo)
      })
    }
  }

  // @override
  selectFile(path) {
    // console.log('select file: ' + path)
    const $pjaxContainer = $(OSC_PJAX_CONTAINER_SEL)
    super.selectFile(path, { '$pjaxContainer': $pjaxContainer, fragment: OSC_PJAX_CONTAINER_SEL })
  }

  // @override
  loadCodeTree(opts, cb) {
    opts.encodedBranch = encodeURIComponent(decodeURIComponent(opts.repo.branch))
    opts.path = (opts.node && (opts.node.sha || opts.encodedBranch)) ||
      (opts.encodedBranch + '&recursive=true')
    this._loadCodeTreeInternal(opts, null, cb)
  }

  // @override
  _getTree(path, opts, cb) {
    this._get(`/tree?ref=${path}`, opts, (err, res) => {
      if (err) cb(err)
      else cb(null, res)
    })
  }

  // @override
  _getSubmodules(tree, opts, cb) {
    cb()
  // const item = tree.filter((item) => /^\.gitmodules$/i.test(item.path))[0]
  // if (!item) return cb()
  // this._get(`/git/blobs/${item.sha}`, opts, (err, res) => {
  //     if (err) return cb(err)
  //     const data = atob(res.content.replace(/\n/g, ''))
  //     cb(null, parseGitmodules(data))
  // })
  }

  _get(path, opts, cb) {
    let array = []
    const host = location.protocol + '//' + location.host
    const id = $('#search_project_id').val().trim()

    var request = (page) => {

      var url = `${host}/api/v4/projects/${id}/repository${path || ''}`
      if (opts.token) {
        url += (url.indexOf("?") >= 0 ? "&" : "?") + `private_token=${opts.token}`
      }
      url += '&page='+page
      const cfg = {
        url,
        method: 'GET',
        cache: false,
        xhrFields: {
          withCredentials: true
        },
      }

      $.ajax(cfg)
        .done((data, textStatus, jqXHR) => {
          if (path && path.indexOf('/git/trees') === 0 && data.truncated) {
            this._handleError({ status: 206 }, cb)
          }
          const totalPages  = parseInt(jqXHR.getResponseHeader('x-total-pages'),10)
          array = array.concat(data)
          if (page < totalPages){
            request(page+1)
          }else {
            cb(null, array)
          }
        })
        .fail((jqXHR) => {
          if (retry) {
            request(false)
          } else {
            this._handleError(jqXHR, cb)
          }
        })
    }
    request(1)
  }
}


src/adapters/adapter.js
class Adapter {
  constructor(deps, store) {
    deps.forEach(dep => window[dep]())
    this._defaultBranch = {}
    this.store = store
  }

  /**
   * Loads the code tree of a repository.
   * @param {Object} opts: {
   *                  path: the starting path to load the tree,
   *                  repo: the current repository,
   *                  node (optional): the selected node (null to load entire tree),
   *                  token (optional): the personal access token
   *                 }
   * @param {Function} transform(item)
   * @param {Function} cb(err: error, tree: Array[Array|item])
   * @api protected
   */
  _loadCodeTreeInternal(opts, transform, cb) {
    const folders = { '': [] }
    const $dummyDiv = $('<div/>')
    const {path, repo, node} = opts

    opts.encodedBranch = opts.encodedBranch || encodeURIComponent(decodeURIComponent(repo.branch))

    this._getTree(path, opts, (err, tree) => {
      if (err) return cb(err)

      this._getSubmodules(tree, opts, (err, submodules) => {
        if (err) return cb(err)

        submodules = submodules || {}

        const nextChunk = (iteration = 0) => {
          const CHUNK_SIZE = 300

          for (let i = 0; i < CHUNK_SIZE; i++) {
            const item = tree[iteration * CHUNK_SIZE + i]

            // we're done
            if (item === undefined) {
              return cb(null, folders[''])
            }

            // runs transform requested by subclass
            if (transform) {
              transform(item)
            }

            // if lazy load and has parent, prefix with parent path
            if (node && node.path) {
              item.path = node.path + '/' + item.path
            }

            const path = item.path
            const type = item.type
            const index = path.lastIndexOf('/')
            const name = $dummyDiv.text(path.substring(index + 1)).html() // sanitizes, closes #9

            item.id = NODE_PREFIX + path
            item.text = name

            // uses `type` as class name for tree node
            item.icon = type

            // @ifdef SUPPORT_FILE_ICONS
            if (type === 'blob') {
              if (this.store.get(STORE.ICONS)) {
                const className = FileIcons.getClassWithColor(name)
                item.icon += ' ' + (className || 'file-generic')
              }
              else {
                item.icon += ' file-generic'
              }
            }
            // @endif

            // @ifndef SUPPORT_FILE_ICONS
            item.icon += ' file-generic'
            // @endif

            if (item.patch) {
              let patch_html = ''

              switch (item.patch.action) {
                case 'added':
                  patch_html += '<span class="text-green">added</span>'
                  break
                case 'renamed':
                  patch_html +=
                    `<span class="text-green" title="${item.patch.previous}">renamed</span>`
                  break
                case 'removed':
                  patch_html +=
                    `<span class="text-red" title="${item.patch.previous}">removed</span>`
                  break
                default:
                  break
              }

              if (item.patch.filesChanged) {
                const fileString = item.patch.filesChanged === 1 ? 'file' : 'files'
                patch_html += `<span>${item.patch.filesChanged} ${fileString}</span>`
              }

              if (item.patch.additions !== 0) {
                patch_html += `<span class="text-green">+${item.patch.additions}</span>`
              }
              if (item.patch.deletions !== 0) {
                patch_html += `<span class="text-red">-${item.patch.deletions}</span>`
              }

              item.text += `<span class="patch">${patch_html}</span>`
            }

            if (node) {
              folders[''].push(item)
            } else {
              folders[path.substring(0, index)].push(item)
            }

            if (type === 'tree' || type === 'blob') {
              if (type === 'tree') {
                if (node)
                  item.children = true
                else
                  folders[item.path] = item.children = []
              }

              // if item is part of a PR, jump to that file's diff
              if (item.patch && typeof item.patch.diffId === 'number') {
                const url = this._getPatchHref(repo, item.patch)
                item.a_attr = {
                  href: url,
                  'data-download-url': item.url,
                  'data-download-filename': name,
                }
              } else {
                // encodes but retains the slashes, see #274
                const encodedPath = path.split('/').map(encodeURIComponent).join('/')
                const url = this._getItemHref(repo, type, encodedPath, opts.encodedBranch)
                item.a_attr = {
                  href: url,
                  'data-download-url': url,
                  'data-download-filename': name,
                }
              }
            } else if (type === 'commit') {
              let moduleUrl = submodules[item.path]
              if (moduleUrl) { // fixes #105
                // special handling for submodules hosted in GitHub
                if (~moduleUrl.indexOf('github.com')) {
                  moduleUrl = moduleUrl.replace(/^git(:\/\/|@)/, window.location.protocol + '//')
                    .replace('github.com:', 'github.com/')
                    .replace(/.git$/, '')
                  item.text = `<a href="${moduleUrl}" class="jstree-anchor">${name}</a>
                 <span>@ </span>
                 <a href="${moduleUrl}/tree/${item.sha}" class="jstree-anchor">${item.sha.substr(0, 7)}</a>`
                }
                item.a_attr = { href: moduleUrl }
              }
            }
          }

          setTimeout(() => nextChunk(iteration + 1))
        }

        nextChunk()
      })
    })
  }

  /**
   * Generic error handler.
   * @api protected
   */
  _handleError(jqXHR, cb) {
    let error, message, needAuth

    switch (jqXHR.status) {
    case 0:
      error = '连接错误'
      message =
        `无法连接到网站. 如果你的网络连接这个网站很好,也许有一个中断的API. 请稍后再试.`
      needAuth = false
      break
    case 206:
      error = '仓库太大'
      message =
        `这个仓库检索太大. 如果你经常使用这个库,去设置和取消“立即加载整个仓库”的选项.`
      break
    case 401:
      error = '无效的token'
      message =
        `token是无效的.<br/>
      <a href="${this.getCreateTokenUrl()}" target="_blank">点此</a>
       去创建一个access token并粘贴到下面.`
      needAuth = true
      break
    case 409:
      error = '空仓库'
      message = '空仓库.'
      break
    case 404:
      error = '私人仓库'
      message =
        `访问私有仓库需要access token.<br/>
       <a href="${this.getCreateTokenUrl()}" target="_blank">点此链接</a>
       去创建一个access token并粘贴到下面.`
      needAuth = true
      break
    case 403:
      if (~jqXHR.getAllResponseHeaders().indexOf('X-RateLimit-Remaining: 0')) {
        // It's kinda specific for GitHub
        error = 'API超过限制'
        message =
          `你已经超过GitHub API小时限制和需要GitHub访问令牌进行额外的请求.<br/>
          <a href="${this.getCreateTokenUrl()}" target="_blank">点此</a>
           去创建一个access token并粘贴到下面.`
        needAuth = true
        break
      } else {
        error = '禁止访问'
        message =
          `禁止访问.
         你可能需要提供 access token.<br/>
       <a href="${this.getCreateTokenUrl()}" target="_blank">点此链接</a>
       去创建一个access token并粘贴到下面.`
        needAuth = true
        break
      }
    default:
      error = message = jqXHR.statusText
      needAuth = false
      break
    }
    cb({
      error: `错误信息: ${error}`,
      message: message,
      needAuth: needAuth
    })
  }

  /**
   * Returns the CSS class to be added to the Octotree sidebar.
   * @api protected
   */
  _getCssClass() {
    throw new Error('没有实现的方法')
  }

  /**
   * Returns the minimum width acceptable for the sidebar.
   * @api protected
   */
  _getMinWidth() {
    return 200
  }

  /**
   * Inits behaviors after the sidebar is added to the DOM.
   * @api public
   */
  init($sidebar) {
    $sidebar
      .resizable({ handles: 'e', minWidth: this._getMinWidth() })
      .addClass(this._getCssClass())
  }

  /**
   * Returns whether the adapter is capable of loading the entire tree in
   * a single request. This is usually determined by the underlying the API.
   * @api public
   */
  canLoadEntireTree() {
    return false
  }

  /**
   * Loads the code tree.
   * @api public
   */
  loadCodeTree(opts, cb) {
    throw new Error('没有实现的方法')
  }

  /**
   * Returns the URL to create a personal access token.
   * @api public
   */
  getCreateTokenUrl() {
    throw new Error('没有实现的方法')
  }

  /**
   * Updates the layout based on sidebar visibility and width.
   * @api public
   */
  updateLayout(togglerVisible, sidebarVisible, sidebarWidth) {
    throw new Error('没有实现的方法')
  }

  /**
   * Returns repo info at the current path.
   * @api public
   */
  getRepoFromPath(token, cb) {
    throw new Error('没有实现的方法')
  }

  /**
   * Selects the file at a specific path.
   * @api public
   */
  selectFile(path) {
    window.location.href = path
  }

  /**
   * Selects a submodule.
   * @api public
   */
  selectSubmodule(path) {
    window.location.href = path
  }

  /**
   * Opens file or submodule in a new tab.
   * @api public
   */
  openInNewTab(path) {
    window.open(path, '_blank').focus()
  }

  /**
   * Downloads a file.
   * @api public
   */
  downloadFile(path, fileName) {
    const link = document.createElement('a')
    link.setAttribute('href', path.replace(/\/blob\/|\/src\//, '/raw/'))
    link.setAttribute('download', fileName)
    link.click()
  }

  /**
   * Gets tree at path.
   * @param {Object} opts - {token, repo}
   * @api protected
   */
  _getTree(path, opts, cb) {
    throw new Error('没有实现的方法')
  }

  /**
   * Gets submodules in the tree.
   * @param {Object} opts - {token, repo, encodedBranch}
   * @api protected
   */
  _getSubmodules(tree, opts, cb) {
    throw new Error('没有实现的方法')
  }

  /**
   * Returns item's href value.
   * @api protected
   */
   _getItemHref(repo, type, encodedPath, encodedBranch) {
     return `/${repo.username}/${repo.reponame}/${type}/${encodedBranch}/${encodedPath}`
   }
   /**
    * Returns patch's href value.
    * @api protected
    */
   _getPatchHref(repo, patch) {
     return `/${repo.username}/${repo.reponame}/pull/${repo.pullNumber}/files#diff-${patch.diffId}`
   }
 }
